File: essentials/chapter_7/agent_tools.py
"""Agent Tools - LangChain with Wikipedia."""

from __future__ import annotations

import json
import os
import re
import sys
from typing import Dict, List, Optional, Tuple

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))

sys.path.append(REPO_ROOT)

if any(arg in {"-h", "--help"} for arg in sys.argv[1:]):
    from shared.utils import print_cli_help

    print_cli_help(sys.argv[0])
    sys.exit(0)


from essentials.chapter_6.agent_memory_retrieval import LangChainLLMManager as Chapter6LangChainManager
from essentials.chapter_7.tools import build_tools_prompt, run_tool
from shared.essentials.utils import interactive_cli

TOOLS_TEMPLATE = """
You are a helpful assistant with access to external tools.

Available tools:
{tools}

Return strict JSON:
{{
  "tool_calls": [{{"name": "tool_name", "arguments": {{"arg": "value"}}, "output": null}}],
  "final_answer": "string"
}}

Rules:
- DEFAULT BEHAVIOR: For most user queries, first call "get_wikipedia_evidence_pack" with arguments {{"query": <user topic>}}.
  This is especially important for topics involving factual claims, names, dates, definitions, history, science, places, people, events, or "what is/why/how" questions.
- Use an empty tool_calls array only if the user request is purely creative writing, brainstorming fiction, or personal preference with no need for sources.
- If you call a tool, include it in tool_calls with output set to null and keep final_answer short until you see the tool output.
- Return JSON only.

User topic: {topic}
""".strip()


FOLLOW_UP_TEMPLATE = """
You already requested a tool and now have the result.

Original user topic: {topic}
Tool call: {tool_call}
Tool output: {tool_output}

Return strict JSON:
{{
  "tool_calls": [{{"name": "tool_name", "arguments": {{"arg": "value"}}, "output": "serialized tool output"}}],
  "final_answer": "final response for the user"
}}
""".strip()


class LangChainLLMManager(Chapter6LangChainManager):
    """Chapter 7 LangChain manager with retrieval-aware tool orchestration."""

    def __init__(self, memory_enabled: bool = True, retrieval_k: int = 4):
        super().__init__(memory_enabled=memory_enabled, retrieval_k=retrieval_k)
        self.framework = "LangChain Tools"

    @staticmethod
    def _extract_json_object(raw: str) -> Dict:
        text = raw.strip()
        if text.startswith("```json"):
            text = text[7:]
        if text.startswith("```"):
            text = text[3:]
        if text.endswith("```"):
            text = text[:-3]

        text = text.strip()
        try:
            return json.loads(text)
        except json.JSONDecodeError:
            match = re.search(r"\{[\s\S]*\}", text)
            if not match:
                raise
            return json.loads(match.group(0))

    def _invoke_json_step(self, provider: str, prompt: str, temperature: float, max_tokens: int) -> Tuple[Dict, object]:
        model = self._create_model(provider, temperature=temperature, max_tokens=max_tokens)
        result = model.invoke(self._build_messages(prompt))
        text = self._extract_text(provider, result)
        payload = self._extract_json_object(text)
        tool_calls = payload.get("tool_calls")
        if not isinstance(tool_calls, list):
            payload["tool_calls"] = []
        return payload, result

    def _load_retrieval_messages(self, session_id: str) -> List:
        """Load persisted history using Chapter 6's session-only memory contract."""
        return self._get_history(session_id).messages

    def _build_retrieval_context(self, topic: str, session_id: str) -> Tuple[str, Dict[str, object]]:
        messages: List = self._load_retrieval_messages(session_id) if self.retrieval_memory_enabled else []
        retrieved = self._select_retrieved_messages(topic, messages) if self.retrieval_memory_enabled else []

        retrieved_context = "\n".join(f"[{item['role']}] {item['content']}" for item in retrieved)
        retrieval_augmented_topic = (
            f"Relevant memory snippets:\n{retrieved_context}\n\nCurrent user topic: {topic}" if retrieved_context else topic
        )

        full_history_context = "\n".join(
            f"[{getattr(msg, 'type', 'unknown')}] {getattr(msg, 'content', '')}" for msg in messages
        )
        prompt_without_retrieval = (
            f"{full_history_context}\n\nCurrent user topic: {topic}" if full_history_context else topic
        )

        tokens_with_retrieval = self._estimate_tokens(retrieval_augmented_topic)
        tokens_without_retrieval = self._estimate_tokens(prompt_without_retrieval)
        estimated_saved = max(0, tokens_without_retrieval - tokens_with_retrieval)
        reduction_percent = round((estimated_saved / tokens_without_retrieval) * 100, 2) if tokens_without_retrieval else 0.0

        retrieval_metadata = {
            "history_messages_available": len(messages),
            "retrieved_messages_count": len(retrieved),
            "retrieved_messages": retrieved,
            "tokens_with_memory_retrieval": tokens_with_retrieval,
            "tokens_without_memory_retrieval": tokens_without_retrieval,
            "estimated_tokens_saved": estimated_saved,
            "estimated_token_reduction_percent": reduction_percent,
        }
        return retrieval_augmented_topic, retrieval_metadata


    def _safe_reset_corrupt_history(self, session_id: str) -> None:
        session_path = self._session_file_path(session_id)
        if not session_path.exists():
            return

        backup_path = session_path.with_suffix(f"{session_path.suffix}.corrupt")
        try:
            if backup_path.exists():
                backup_path.unlink()
            session_path.replace(backup_path)
        except Exception:
            # If backup move fails, continue by overwriting in place.
            pass

        session_path.write_text("[]", encoding="utf-8")

    def _get_history(self, session_id: str):
        history = super()._get_history(session_id)
        try:
            _ = history.messages
            return history
        except Exception as exc:
            message = str(exc)
            known_corruption_signals = (
                "string indices must be integers",
                "Expecting value",
                "JSON",
            )
            if not any(signal in message for signal in known_corruption_signals):
                raise

            self.histories.pop(session_id, None)
            self._safe_reset_corrupt_history(session_id)
            repaired_history = super()._get_history(session_id)
            _ = repaired_history.messages
            return repaired_history

    @staticmethod
    def _resolve_tools_template(template: Optional[str]) -> str:
        candidate = (template or "").strip()
        if not candidate or candidate == "{topic}" or "{tools}" not in candidate:
            return TOOLS_TEMPLATE
        return template

    def ask_question(
        self,
        topic: str,
        provider: Optional[str] = None,
        template: str = TOOLS_TEMPLATE,
        max_tokens: int = 1000,
        temperature: float = 0.2,
        session_id: str = "default",
    ) -> Dict:
        template = self._resolve_tools_template(template)
        provider = self.resolve_model_identifier(provider)
        prompt = template.format(topic=topic, tools=build_tools_prompt())

        if not provider:
            return {
                "success": False,
                "error": "No providers available",
                "provider": "none",
                "model": "none",
                "prompt": prompt,
                "response": None,
            }

        model_config = self.resolve_model_config(provider)
        model = model_config.model
        retrieval_topic, retrieval_metadata = self._build_retrieval_context(topic, session_id)
        retrieval_prompt = template.format(topic=retrieval_topic, tools=build_tools_prompt())

        try:
            first_step, first_result = self._invoke_json_step(provider, retrieval_prompt, temperature=temperature, max_tokens=max_tokens)
            tool_calls = first_step.get("tool_calls") or []
            final_answer = str(first_step.get("final_answer", "")).strip()

            executed_tool_calls = [dict(tool_call) for tool_call in tool_calls]
            if executed_tool_calls:
                for tool_call in executed_tool_calls:
                    tool_output = run_tool(tool_call["name"], tool_call.get("arguments") or {})
                    tool_call["output"] = tool_output

                follow_up_prompt = FOLLOW_UP_TEMPLATE.format(
                    topic=topic,
                    tool_call=json.dumps(executed_tool_calls, ensure_ascii=False),
                    tool_output=json.dumps(executed_tool_calls, ensure_ascii=False),
                )
                second_step, _ = self._invoke_json_step(provider, follow_up_prompt, temperature=temperature, max_tokens=max_tokens)
                final_answer = str(second_step.get("final_answer", final_answer)).strip() or final_answer

            raw_response = json.dumps(
                {
                    "tool_calls": executed_tool_calls,
                    "final_answer": final_answer,
                },
                ensure_ascii=False,
            )

            response_metadata = getattr(first_result, "response_metadata", None)
            usage_metadata = getattr(first_result, "usage_metadata", None)
            token_usage = self._extract_token_usage(response_metadata, usage_metadata)

            payload = {
                "tool_calls": executed_tool_calls,
                "final_answer": final_answer,
                "metadata": {
                    **self._build_metadata(
                        {
                            "provider": model_config.provider,
                            "model": model_config.model,
                            "model_identifier": model_config.name,
                            "session_id": session_id,
                            "temperature": temperature,
                            "max_tokens": max_tokens,
                            "response_metadata": response_metadata,
                            "usage_metadata": usage_metadata,
                        },
                        raw_response,
                    ),
                    "retrieval": retrieval_metadata,
                },
            }

            if self.retrieval_memory_enabled:
                history = self._get_history(session_id)
                history.add_user_message(topic)
                history.add_ai_message(raw_response)

            result = {
                "success": True,
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "prompt": retrieval_prompt,
                "response": payload,
                "raw_answer": final_answer,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
                "response_metadata": response_metadata,
                "usage_metadata": usage_metadata,
                "token_usage": token_usage,
            }
            return {k: v for k, v in result.items() if v is not None}
        except Exception as exc:
            return {
                "success": False,
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "prompt": retrieval_prompt,
                "error": str(exc),
                "response": None,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
            }


def main():
    args = sys.argv[1:]
    if "web" in args:
        from shared.essentials.web import run_web_server

        run_web_server(lambda: LangChainLLMManager(memory_enabled=True))
    else:
        interactive_cli(LangChainLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()

File: essentials/chapter_7/agent_tools.ts
/**
 * Agent Tools - LangChain TypeScript with Wikipedia.
 */

import { LangChainLLMManager as Chapter6LangChainManager } from '../../chapter_6/agent_memory_retrieval.ts';
import { interactiveCli, normalizeResponseText, parseStructuredJsonResponse, printCliHelp } from '../../../shared/essentials/utils.ts';
import { buildToolsPrompt, runTool } from '../tools.ts';

const TOOLS_TEMPLATE = `You are a helpful assistant with access to external tools.

Available tools:
{tools}

Return strict JSON:
{
  "tool_calls": [{"name": "tool_name", "arguments": {"arg": "value"}, "output": null}],
  "final_answer": "string"
}

Rules:
- DEFAULT BEHAVIOR: For most user queries, first call "get_wikipedia_evidence_pack" with arguments {{"query": <user topic>}}.
  This is especially important for topics involving factual claims, names, dates, definitions, history, science, places, people, events, or "what is/why/how" questions.
- Use an empty tool_calls array only if the user request is purely creative writing, brainstorming fiction, or personal preference with no need for sources.
- If you call a tool, include it in tool_calls with output set to null and keep final_answer short until you see the tool output.
- Return JSON only.

User topic: {topic}`;

const FOLLOW_UP_TEMPLATE = `You already requested a tool and now have the result.

Original user topic: {topic}
Tool call: {tool_call}
Tool output: {tool_output}

Return strict JSON:
{
  "tool_calls": [{"name": "tool_name", "arguments": {"arg": "value"}, "output": "serialized tool output"}],
  "final_answer": "final response for the user"
}`;

class LangChainLLMManager extends Chapter6LangChainManager {
    constructor(memoryEnabled = true, retrievalK = 4) {
        super(memoryEnabled, retrievalK);
        this.framework = 'LangChain Tools TypeScript';
    }

    _extractJsonObject(raw) {
        return parseStructuredJsonResponse(raw);
    }

    _buildFallbackToolPayload(rawText) {
        const fallbackAnswer = normalizeResponseText(rawText).trim();
        if (!fallbackAnswer) {
            throw new Error('No JSON object found in model response');
        }
        return {
            tool_calls: [],
            final_answer: fallbackAnswer,
        };
    }

    _normalizeToolPayload(rawPayload, rawText) {
        if (!rawPayload || typeof rawPayload !== 'object' || Array.isArray(rawPayload)) {
            return this._buildFallbackToolPayload(rawText);
        }

        const toolCalls = Array.isArray(rawPayload.tool_calls) ? rawPayload.tool_calls : [];
        const finalAnswer = rawPayload.final_answer;

        if (typeof finalAnswer === 'string' && finalAnswer.trim()) {
            return {
                tool_calls: toolCalls
                    .filter((toolCall) => toolCall && typeof toolCall === 'object' && toolCall.name)
                    .map((toolCall) => ({
                        name: String(toolCall.name),
                        arguments: toolCall.arguments && typeof toolCall.arguments === 'object' ? toolCall.arguments : {},
                        ...(Object.prototype.hasOwnProperty.call(toolCall, 'output') ? { output: toolCall.output } : {}),
                    })),
                final_answer: finalAnswer.trim(),
            };
        }

        return this._buildFallbackToolPayload(rawText);
    }

    async _invokeJsonStep(provider, prompt, temperature, maxTokens) {
        const model = this._createModel(provider, temperature, maxTokens);
        const result = await model.invoke(this._buildMessages(prompt));
        const text = this._extractText(provider, result);

        try {
            const payload = this._extractJsonObject(text);
            return { payload: this._normalizeToolPayload(payload, text), result };
        } catch {
            return { payload: this._buildFallbackToolPayload(text), result };
        }
    }

    async _loadRetrievalMessages(sessionId) {
        const history = this._getHistory(sessionId);
        return history.getMessages();
    }

    async _buildRetrievalContext(topic, sessionId) {
        const messages = this.retrievalMemoryEnabled
            ? await this._loadRetrievalMessages(sessionId)
            : [];

        const retrievedResult = this.retrievalMemoryEnabled ? await this._selectRetrievedMessages(topic, messages) : [];
        const retrieved = Array.isArray(retrievedResult) ? retrievedResult : [];
        const retrievedContext = retrieved.map((item) => `[${item.role}] ${item.content}`).join('\n');
        const retrievalAugmentedTopic = retrievedContext
            ? `Relevant memory snippets:\n${retrievedContext}\n\nCurrent user topic: ${topic}`
            : topic;

        const fullHistoryContext = messages
            .map((msg) => `[${msg?._getType?.() ?? msg?.getType?.() ?? msg?.type ?? 'unknown'}] ${String(msg?.content ?? '')}`)
            .join('\n');
        const promptWithoutRetrieval = fullHistoryContext
            ? `${fullHistoryContext}\n\nCurrent user topic: ${topic}`
            : topic;

        const tokensWithRetrieval = this._estimateTokens(retrievalAugmentedTopic);
        const tokensWithoutRetrieval = this._estimateTokens(promptWithoutRetrieval);
        const estimatedSaved = Math.max(0, tokensWithoutRetrieval - tokensWithRetrieval);
        const reductionPercent = tokensWithoutRetrieval > 0
            ? Number(((estimatedSaved / tokensWithoutRetrieval) * 100).toFixed(2))
            : 0;

        return {
            retrievalAugmentedTopic,
            retrievalMetadata: {
                history_messages_available: messages.length,
                retrieved_messages_count: retrieved.length,
                retrieved_messages: retrieved,
                tokens_with_memory_retrieval: tokensWithRetrieval,
                tokens_without_memory_retrieval: tokensWithoutRetrieval,
                estimated_tokens_saved: estimatedSaved,
                estimated_token_reduction_percent: reductionPercent,
            },
        };
    }



    _resolveToolsTemplate(template) {
        const candidate = String(template ?? '').trim();
        if (!candidate || candidate === '{topic}' || !candidate.includes('{tools}')) {
            return TOOLS_TEMPLATE;
        }
        return template;
    }

    _normalizeWikipediaQuery(topic) {
        const text = String(topic || '').trim();
        if (!text) return text;
        return text
            .replace(/^\s*(what\s+is|who\s+is|where\s+is|when\s+did|when\s+was|why\s+is|how\s+is)\s+/i, '')
            .replace(/[?]+$/g, '')
            .trim();
    }

    _shouldForceWikipediaTool(topic, toolCall) {
        if (toolCall && typeof toolCall === 'object' && toolCall.name) return false;
        const text = String(topic || '').toLowerCase();
        if (!text.trim()) return false;

        const creativeSignals = [
            'poem', 'story', 'fiction', 'brainstorm', 'imagine', 'creative writing', 'roleplay', 'joke',
        ];
        if (creativeSignals.some((k) => text.includes(k))) return false;

        const factualSignals = [
            'what is', 'who is', 'when did', 'where is', 'why', 'how', 'define', 'history', 'date', 'science',
        ];
        return factualSignals.some((k) => text.includes(k)) || text.split(/\s+/).length >= 3;
    }

    async askQuestion(topic, provider = null, template = TOOLS_TEMPLATE, maxTokens = 1000, temperature = 0.2, sessionId = 'default') {
        template = this._resolveToolsTemplate(template);
        const resolvedProvider = this.resolveModelIdentifier(provider);
        const basePrompt = template.replace('{topic}', topic).replace('{tools}', buildToolsPrompt());

        if (!resolvedProvider) {
            return {
                success: false,
                error: 'No providers available',
                provider: 'none',
                model: 'none',
                prompt: basePrompt,
                response: null,
            };
        }

        const modelConfig = this.resolveModelConfig(resolvedProvider);

        const { retrievalAugmentedTopic, retrievalMetadata } = await this._buildRetrievalContext(topic, sessionId);
        const prompt = template.replace('{topic}', retrievalAugmentedTopic).replace('{tools}', buildToolsPrompt());
        try {
            const { payload: firstStep, result: firstResult } = await this._invokeJsonStep(resolvedProvider, prompt, temperature, maxTokens);
            const toolCalls = Array.isArray(firstStep.tool_calls) ? firstStep.tool_calls : [];
            let finalAnswer = String(firstStep.final_answer || '').trim();
            let executedToolCalls = toolCalls.map((toolCall) => ({ ...toolCall }));
            if (this._shouldForceWikipediaTool(topic, executedToolCalls[0])) {
                executedToolCalls = [{
                    name: 'get_wikipedia_evidence_pack',
                    arguments: { query: this._normalizeWikipediaQuery(topic) || topic },
                }];
            }

            if (executedToolCalls.length > 0) {
                for (const toolCall of executedToolCalls) {
                    const toolName = String(toolCall.name);
                    const toolArgs = toolCall.arguments && typeof toolCall.arguments === 'object' ? toolCall.arguments : {};
                    toolCall.output = await runTool(toolName, toolArgs);
                }

                const followUpPrompt = FOLLOW_UP_TEMPLATE
                    .replace('{topic}', topic)
                    .replace('{tool_call}', JSON.stringify(executedToolCalls))
                    .replace('{tool_output}', JSON.stringify(executedToolCalls));

                const { payload: secondStep } = await this._invokeJsonStep(resolvedProvider, followUpPrompt, temperature, maxTokens);
                finalAnswer = String(secondStep.final_answer || finalAnswer).trim() || finalAnswer;
            }

            const rawResponse = JSON.stringify({
                tool_calls: executedToolCalls,
                final_answer: finalAnswer,
            });

            const responseMetadata = firstResult?.response_metadata ?? firstResult?.responseMetadata ?? null;
            const usageMetadata = firstResult?.usage_metadata ?? firstResult?.usageMetadata ?? null;
            const tokenUsage = this._extractTokenUsage(responseMetadata, usageMetadata);

            const responsePayload = {
                tool_calls: executedToolCalls,
                final_answer: finalAnswer,
                metadata: {
                    ...this._buildMetadata({
                        provider: modelConfig.provider,
                        model: modelConfig.model,
                        modelIdentifier: modelConfig.name,
                        sessionId,
                        temperature,
                        maxTokens,
                        response_metadata: responseMetadata,
                        usage_metadata: usageMetadata,
                    }, rawResponse),
                    retrieval: retrievalMetadata,
                },
            };

            if (this.retrievalMemoryEnabled) {
                const history = this._getHistory(sessionId);
                await history.addUserMessage(topic);
                await history.addAIMessage(rawResponse);
            }

            return {
                success: true,
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                prompt,
                response: responsePayload,
                rawAnswer: finalAnswer,
                temperature,
                maxTokens,
                sessionId,
                ...(responseMetadata != null ? { response_metadata: responseMetadata } : {}),
                ...(usageMetadata != null ? { usage_metadata: usageMetadata } : {}),
                ...(tokenUsage != null ? { token_usage: tokenUsage } : {}),
            };
        } catch (error) {
            return {
                success: false,
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                prompt,
                error: error.message,
                response: null,
                temperature,
                maxTokens,
                sessionId,
            };
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.includes('-h') || args.includes('--help')) {
        printCliHelp(process.argv[1]);
        return;
    }
    if (args.includes('web')) {
        const { runWebServer } = await import('../../../shared/essentials/web.ts');
        await runWebServer(() => new LangChainLLMManager(true));
    } else {
        const manager = new LangChainLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LangChainLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}
